internal/consensus/state.go

```diff
@@ -1620,30 +1620,69 @@ func (cs *State) defaultDoPrevote(ctx context.Context, height int64, round int32
 		return
 	}
 
-	// If we're not the proposer, we need to build the block
-	if cs.config.GossipTransactionKeyOnly && cs.Proposal != nil && cs.ProposalBlock == nil {
-		txKeys := cs.Proposal.TxKeys
-		if cs.ProposalBlockParts.IsComplete() {
+	if cs.config.GossipTransactionKeyOnly {
+		if cs.ProposalBlock == nil {
+			// If we're not the proposer, we need to build the block
+			txKeys := cs.Proposal.TxKeys
+			if cs.ProposalBlockParts.IsComplete() {
+				block, err := cs.getBlockFromBlockParts()
+				if err != nil {
+					cs.logger.Error("Encountered error building block from parts", "block parts", cs.ProposalBlockParts)
+					return
+				}
+				// We have full proposal block and txs. Build proposal block with txKeys
+				proposalBlock := cs.buildProposalBlock(height, block.Header, block.LastCommit, block.Evidence, block.ProposerAddress, txKeys)
+				if proposalBlock == nil {
+					return
+				}
+				cs.ProposalBlock = proposalBlock
+			} else {
+				return
+			}
+		}
+	} else {
+		if cs.ProposalBlock == nil {
 			block, err := cs.getBlockFromBlockParts()
 			if err != nil {
 				cs.logger.Error("Encountered error building block from parts", "block parts", cs.ProposalBlockParts)
 				return
 			}
-			// We have full proposal block and txs. Build proposal block with txKeys
-			proposalBlock := cs.buildProposalBlock(height, block.Header, block.LastCommit, block.Evidence, block.ProposerAddress, txKeys)
-			if proposalBlock == nil {
+			if block == nil {
+				logger.Debug("prevote step: ProposalBlock is nil")
+				cs.signAddVote(ctx, tmproto.PrevoteType, nil, types.PartSetHeader{})
 				return
 			}
-			cs.ProposalBlock = proposalBlock
-		} else {
+			cs.ProposalBlock = block
+		}
+		txKeys := cs.Proposal.TxKeys
+		// add missing tx keys to the mempool for CheckTx
+		missingTxKeys := cs.blockExec.GetMissingTxs(txKeys)
+		blockTxKeyToTx := map[types.TxKey]types.Tx{}
+		for _, tx := range cs.ProposalBlock.Txs {
+			blockTxKeyToTx[tx.Key()] = tx
+		}
+		for _, missingTxKey := range missingTxKeys {
```

Comment:
Any chance of this causing a race condition between different validator nodes?

Reply:
this part is likely to run differently for different validators, which is kind of the point of doing it here, since different validators will have different min gas app configs so some txs might be accepted by some vals but rejected by others, and we want those who would reject those txs to reject (i.e. vote nil) its enclosing block right here. Since this isn't part of block processing, difference in validator behaviors should be fine